packages/devtools_app/lib/src/flutter/console.dart

```diff
@@ -0,0 +1,219 @@
+
+  // Whenever the widget updates, refresh the scroll position if needed.
+  @override
+  void didUpdateWidget(oldWidget) {
+    if (_scroll.hasClients && _scroll.atScrollBottom) {
+      _scroll.autoScrollToBottom();
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final theme = Theme.of(context);
+    final textStyle =
+        theme.textTheme.bodyText2.copyWith(fontFamily: 'RobotoMono');
+
+    return OutlineDecoration(
+      child: Scrollbar(
+        child: ListView.builder(
+          padding: const EdgeInsets.all(denseSpacing),
+          itemCount: widget.lines?.length ?? 0,
+          itemExtent: CodeView.rowHeight, // TODO: Get from theme?
+          controller: _scroll,
+          itemBuilder: (context, index) {
+            return RichText(
+              text: TextSpan(
+                children: processAnsiTerminalCodes(
+                  widget.lines[index],
+                  textStyle,
+                ),
+              ),
+              maxLines: 1,
+            );
+          },
+        ),
+      ),
+    );
+  }
+}
+
+// CONTROLS
+
+/// A pre-configured IconButton that fits the ux of the Console widget.
+///
+/// The customizations are:
+///  * Icon size: [actionsIconSize]
+///  * Do not show [tooltip] if the button is disabled
+///  * [VisualDensity.compact]
+class ConsoleControl extends StatelessWidget {
+  const ConsoleControl({
+    this.icon,
+    this.tooltip,
+    this.onPressed,
+    this.buttonKey,
+  });
+
+  final IconData icon;
+  final String tooltip;
+  final VoidCallback onPressed;
+  final Key buttonKey;
+
+  @override
+  Widget build(BuildContext context) {
+    final disabled = onPressed == null;
+    return IconButton(
+      icon: Icon(icon, size: actionsIconSize),
+      onPressed: onPressed,
+      tooltip: disabled ? null : tooltip,
+      visualDensity: VisualDensity.compact,
+      key: buttonKey,
+    );
+  }
+}
+
+/// A Console Control to "delete" the contents of the console.
+///
+/// This just preconfigures a ConsoleControl with the `delete` icon,
+/// and the `onPressed` function passed from the outside.
+class DeleteControl extends StatelessWidget {
+  const DeleteControl({
+    this.onPressed,
+    this.tooltip = 'Clear contents',
+    this.buttonKey,
+  });
+
+  final VoidCallback onPressed;
+  final String tooltip;
+  final Key buttonKey;
+
+  @override
+  Widget build(BuildContext context) {
+    return ConsoleControl(
+      icon: Icons.delete,
+      tooltip: tooltip,
+      onPressed: onPressed,
+      buttonKey: buttonKey,
+    );
+  }
+}
+
+/// A Console Control that copies `data` to the clipboard.
+///
+/// If it succeeds, it displays a notification with `successMessage`.
+class CopyToClipboardControl extends StatelessWidget {
+  const CopyToClipboardControl({
+    this.data,
+    this.successMessage = 'Copied to clipboard.',
+    this.tooltip = 'Copy to clipboard',
+    this.buttonKey,
+  });
+
+  final String data;
```

Comment:
This could take a data or a data provider - a callback that could calculate the data.

For an active console, we could be concatenating 5k lines (our current console limit) each time the console widget rebuilt.

Reply:
Good points, I'm going to try and make these computations lazy so they don't impact the main code path.